Add pxe boot

halleelyasaf opened the pull request.

snecklifter committed:
Fixes day2 worker playbook for disconnected installs

Currently the day2 worker playbook is broken for disconnected installs
as the patching logic for mirrored registries does not exist. This
commit fixes this although its mostly a copy of the templates from
the main create cluster role so probably can be done a lot nicer.
Also, the CSR approval mechanism is still broken.

mkochanowski committed:
Add support for Single Node OpenShift deployments

From now on, end users can choose to deploy either a SNO or HA cluster.

SNO will be enabled automatically if and only if one master node and
zero worker nodes are defined in the inventory.

Co-authored-by: Dylan Wong <dywong@redhat.com>
Signed-off-by: Marek Kochanowski <mkochanowski@redhat.com>

nocturnalastro committed:
Remove alias for kvm node hostvars

This causes all values in hostvars to be evaluated, this could cause some values to be evaluated
before the value for them is populated e.g. the file it reads does not exist at that point in
the execution

novacain1 committed:
add boot flag so VM is restarted if host is rebooted

nocturnalastro committed:
Save kubeadmin credentials to a ansible vault

Requires a new variable which kubeadmin_vault_password_file_path
 which defines the path to file containing the password for the vault

mkochanowski committed:
Remove hardcoded value from create_cluster request

This change modifies the POST request of the `create_cluster` role to
reference a `cluster_network_host_prefix` variable instead of using a
hardcoded value of 23 in the payload.

Signed-off-by: Marek Kochanowski <mkochanowski@redhat.com>

halleelyasaf committed:
Add Cleanup before creating the key

Commit:
Add the ability to specify install disk partition for OCP versions greater than 4.8

nocturnalastro committed:
Move REGISTRY_HTTP_SECRET check with other registry var checks

This removes the requirement for registry_host to be defined.

nocturnalastro committed:
Fix issue with nmstate template with port

An issue with the template variable look up which placed the port name
a different level from the rest of the bridge port variables.

halleelyasaf committed:
Add pxe boot

Optionally adds TFTP host and configures dhcp to enable
pxe booting

dependabot[bot] committed:
build(deps): bump actions/setup-python from 2.3.1 to 2.3.2

Bumps [actions/setup-python](https://github.com/actions/setup-python) from 2.3.1 to 2.3.2.
- [Release notes](https://github.com/actions/setup-python/releases)
- [Commits](https://github.com/actions/setup-python/compare/v2.3.1...v2.3.2)

---
updated-dependencies:
- dependency-name: actions/setup-python
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

natifridman commented on playbooks/deploy_httpd_for_pxe.yml:
```diff
@@ -0,0 +1,7 @@
+- name: Setup httpd for pxe
```
```suggestion
- name: Setup httpd for PXE
```

natifridman commented on playbooks/deploy_ipmitool.yml:
```diff
@@ -0,0 +1,7 @@
+- name: Deploy ipmi tools
```
```suggestion
- name: Deploy IPMI tools
```

natifridman commented on roles/boot_iso/defaults/main.yml:
```diff
@@ -2,3 +2,5 @@
 # defaults file for boot_iso
 boot_iso_url: "{{ discovery_iso_server }}/{{ discovery_iso_name }}"
 secure_sushy_tools: "{{ secure | default(true) }}"
+port_number: 82
+doc_root: /var/www/html
```
Maybe change the names to something meaningful

halleelyasaf replied:
this is the old file that I will delete in the future 
the new file is 
https://github.com/halleelyasaf/crucible/blob/add_pxe_boot/roles/setup_pxe_httpd/defaults/main.yml

natifridman commented on roles/boot_iso/tasks/pxe.yml:
```diff
@@ -0,0 +1,119 @@
+---
+
```
Delete empty line

halleelyasaf replied:
This file will be split

natifridman commented on roles/boot_iso/tasks/pxe.yml:
```diff
@@ -0,0 +1,119 @@
+        - httpd
+        - python3-pip
+
+    - name: Install multi python packages with version specifiers
```
Maybe you can name it like
```suggestion
    - name: Install required python packages
```

halleelyasaf replied:
this is the old file that I will delete in the future
the new file is
https://github.com/halleelyasaf/crucible/blob/add_pxe_boot/roles/setup_ipmi_tools/tasks/main.yml

natifridman commented on roles/boot_iso/tasks/pxe.yml:
```diff
@@ -0,0 +1,119 @@
+
+- name: prepare PXE
+  block:
+    - name: ensure the required packgaes installed
```
Capital letter
```suggestion
    - name: Ensure the required packgaes installed
```

natifridman commented on roles/boot_iso/tasks/pxe.yml:
```diff
@@ -0,0 +1,119 @@
+        name:
+          - pyghmi
+
+    - name: "httpd  Configuration File"
```
Maybe you can name it like
```suggestion
    - name: "Create httpd configuration file"
```

natifridman commented on roles/boot_iso/tasks/pxe.yml:
```diff
@@ -0,0 +1,119 @@
+           dest: "/etc/httpd/conf.d/new_conf.conf"
+    
+
+    - name: Start service httpd, if not started
```
```suggestion
    - name: Start httpd service if not started
```

natifridman commented on roles/boot_iso/tasks/pxe.yml:
```diff
@@ -0,0 +1,119 @@
+      with_items:
+        - tftp
+        - httpd
+
```
Delete empty line

natifridman commented on roles/boot_iso/tasks/pxe.yml:
```diff
@@ -0,0 +1,119 @@
+    - name: mount the discovery iso
+      mount:
+        src: "/opt/http_store/data/{{ discovery_iso_name }}"
+        #src: "{{ discovery_iso_server }}/{{ discovery_iso_name }}"
```
I don't think you need both lines, you can uncomment it or remove it

natifridman commented on roles/boot_iso/tasks/pxe.yml:
```diff
@@ -0,0 +1,119 @@
+      vars:
+        tftp_directory: "/var/lib/tftpboot/"
+        mounted_iso_files:
+                #      - /mnt/EFI/redhat/grub.cfg
```
I think you can remove it

natifridman commented on roles/boot_iso/tasks/pxe.yml:
```diff
@@ -0,0 +1,119 @@
+          - /mnt/images/pxeboot/vmlinuz
+          - /mnt/images/pxeboot/initrd.img
+   
+    - name: " grub Configuration File"
```
Maybe you can name it like
```suggestion
    - name: " Create grub configuration file"
```

natifridman commented on roles/boot_iso/tasks/pxe.yml:
```diff
@@ -0,0 +1,119 @@
+    - name: " grub Configuration File"
+      template:
+           src: "./grub.cfg.j2"
+           dest: "/var/lib/tftpboot/grub.cfg"
```
I think you can create a variable for this like you used `tftp_directory`

natifridman commented on roles/boot_iso/tasks/pxe.yml:
```diff
@@ -0,0 +1,119 @@
+    - name: Copy mount rootfs to http directory
+      copy:
+        src: "/mnt/images/pxeboot/rootfs.img"
+        dest: "/var/www/html/"
```
I think you can create a variable for this. Something like `httpd_diredctory`

natifridman commented on roles/boot_iso/tasks/pxe.yml:
```diff
@@ -0,0 +1,119 @@
+
+  become: true
+
+- name: PXE Power off servers
```
```suggestion
- name: Power off bare metal server
```

natifridman commented on roles/boot_iso/tasks/pxe.yml:
```diff
@@ -0,0 +1,119 @@
+    username: "root"
+    password: "calvin"
+
+- name: Set boot for PXE
```
```suggestion
- name: Set next boot for PXE
```

natifridman commented on roles/boot_iso/templates/grub.cfg.j2:
```diff
@@ -0,0 +1,5 @@
+set timeout=1
+menuentry 'Install Red Hat Enterprise Linux CoreOS' --class fedora --class gnu-linux --class gnu --class os {
+	linuxefi vmlinuz random.trust_cpu=on ignition.firstboot ignition.platform.id=metal 'coreos.live.rootfs_url=http://10.19.111.48:82/rootfs.img' 
```
You can use variables here for the URL and port

natifridman commented on roles/setup_ipmi_tools/tasks/main.yml:
```diff
@@ -0,0 +1,18 @@
+---
+- name: Setup IPMI
+  block:
+    - name: Install IPMI
```
```suggestion
    - name: Install IPMI tools
```

natifridman commented on roles/setup_ipmi_tools/tasks/main.yml:
```diff
@@ -0,0 +1,18 @@
+        name: ipmitool
+        state: present
+    
+    - name: Install python3-pip
+      ansible.builtin.package:
+        name: python3-pip
+        state: present
+    
+    - name: Install python for the ipmi tool module
+      pip:
+        name:
+          - pyghmi
```
You already do this in boot_iso

halleelyasaf replied:
yes I moved that from there this is the draft file

natifridman commented on roles/setup_pxe_httpd/tasks/main.yml:
```diff
@@ -0,0 +1,35 @@
+---
+- name: Setup Httpd
```
```suggestion
- name: Setup httpd
```

natifridman commented on roles/setup_pxe_httpd/tasks/main.yml:
```diff
@@ -0,0 +1,35 @@
+---
+- name: Setup Httpd
+  block:
+    - name: Install Httpd
+      ansible.builtin.package:
+        name: httpd
+        state: present
+
+    - name: Configure httpd
+      ansible.builtin.template:
+        src: "./httpd.conf.j2"
+        dest: "/etc/httpd/conf.d/pxe_conf.conf"
+        owner: root
+        group: root
+        mode: 0644
+      notify: Restart httpd
```
You already do this in `boot_iso`

halleelyasaf replied:
yes I moved that from there this is the draft file

natifridman commented on roles/setup_tftp/tasks/main.yml:
```diff
@@ -0,0 +1,23 @@
+---
+- name: Setup TFTP
+  block:
+    - name: Install TFTP
+      ansible.builtin.package:
+        name: tftp-server
+        state: present
+
+
+    - name: Start TFTP
+      ansible.builtin.service:
+        name: tftp.socket 
+        state: started
+        enabled: true
```
You already do this in boot_iso

halleelyasaf replied:
yes I moved that from there this is the draft file

mkochanowski committed:
Fix dnsmasq missing when NetworkManager is used

Signed-off-by: Marek Kochanowski <mkochanowski@redhat.com>

natifridman commented on roles/mount_discovery_iso_for_pxe/defaults/main.yml:
```diff
@@ -0,0 +1,18 @@
+  - "{{mount_directory}}/pxeboot/vmlinuz"
+  - "{{mount_directory}}/pxeboot/initrd.img"
+
+
```
Delete empty line

natifridman commented on roles/mount_discovery_iso_for_pxe/tasks/main.yml:
```diff
@@ -0,0 +1,41 @@
+        state: mounted
+        fstype: iso9660
+        opts: ro,noauto
+        #{{ discovery_iso_server }}
```
Delete comment

natifridman commented on roles/mount_discovery_iso_for_pxe/tasks/main.yml:
```diff
@@ -0,0 +1,41 @@
+        dest: "{{ tftp_directory }}"
+      with_items: "{{ mounted_iso_files }}"
+   
+    - name: " grub Configuration File"
```
```suggestion
    - name: Create grub configuration file
```

natifridman commented on roles/setup_pxe_httpd/tasks/main.yml:
```diff
@@ -0,0 +1,34 @@
+---
+- name: Setup Httpd
+  block:
+    - name: Install Httpd
```
```suggestion
    - name: Install httpd
```

zedr commented on roles/insert_dns_records/templates/openshift-cluster.conf.j2:
```diff
@@ -2,7 +2,7 @@ domain={{ domain }}
 domain-needed
 bogus-priv
-listen-address=127.0.0.1,{{ listen_address }}
+listen-address=127.0.0.1,10.60.0.190
```
Hardcoded IP here. Is this OK?

natifridman replied:
@halleelyasaf Please revert this file and delete the not relevant files

zedr commented on roles/insert_dns_records/templates/openshift-cluster.conf.j2:
```diff
@@ -13,7 +13,7 @@ server={{ upstream_dns }}
 dhcp-option= tag:{{ cluster_name }},option:netmask,{{ (gateway + '/' + prefix | string) | ipaddr('netmask') }}
 dhcp-option= tag:{{ cluster_name }},option:router,{{ gateway }}
-dhcp-option= tag:{{ cluster_name }},option:dns-server,{{ listen_address }}
+dhcp-option= tag:{{ cluster_name }},option:dns-server,10.60.0.190
```
Hardcoded IP here. Is this OK?

abays committed:
Do not gather_facts for unused prereq services

arjuhe committed:
Documentation

Added install bind-utils as it is not part of CentOS8 Appstream.
Changed dig to host as the output is easier to read.

natifridman commented on playbooks/deploy_ipmitool.yml:
```diff
@@ -0,0 +1,7 @@
+- name: Deploy IPMI tools
+  hosts: bastion
+  vars:
+    SETUP_IPMI_TOOLS: "{{setup_pxe_service | default(True)}}"
```
```suggestion
    SETUP_IPMI_TOOLS: "{{ setup_pxe_service | default(True) }}"
```

natifridman commented on playbooks/deploy_tftp.yml:
```diff
@@ -0,0 +1,7 @@
+- name: Setup TFTP
+  hosts: tftp_host
+  vars:
+    SETUP_TFTP_SERVICE: "{{setup_pxe_service | default(True)}}"
```
```suggestion
    SETUP_TFTP_SERVICE: "{{ setup_pxe_service | default(True) }}"
```

natifridman commented on playbooks/mount_discovery_iso_for_pxe.yml:
```diff
@@ -0,0 +1,8 @@
+    - mount_discovery_iso_for_pxe
+  vars:
+    - DOWNLOAD_DEST_FILE: "{{ discovery_iso_name }}"
+    - DOWNLOAD_DEST_PATH: "{{ iso_download_dest_path | default('/opt/http_store/data') }}"
```
Isn't `iso_download_dest_path` saved in the defaults folder? Why do you need default here?
If you need it, maybe save it in a variable?

natifridman commented on roles/boot_iso/tasks/pxe.yml:
```diff
@@ -0,0 +1,37 @@
+  vars:
+    baseuri: "10.19.111.57"
+    username: "root"
+    password: "calvin"
```
Use variables

natifridman commented on roles/boot_iso/tasks/pxe.yml:
```diff
@@ -0,0 +1,37 @@
+    persistent: no
+    uefiboot: yes
+  vars:
+    baseuri: "10.19.111.57"
+    username: "root"
+    password: "calvin"
```
Use variables

halleelyasaf replied:
i will change that to 
```
    url: "{{ base_bmc_address }}"
    user: "{{ hostvars[inventory_hostname]['bmc_user'] }}"
    password: "{{ hostvars[inventory_hostname]['bmc_password'] }}"
```

natifridman commented on roles/boot_iso/tasks/pxe.yml:
```diff
@@ -0,0 +1,37 @@
+  vars:
+    baseuri: "10.19.111.57"
+    username: "root"
+    password: "calvin"
```
Use variables

natifridman commented on roles/insert_dns_records/templates/openshift-cluster.conf.j2:
```diff
@@ -43,3 +43,11 @@ address=/{{ item.address }}/{{ item.ip }}
+
+{% if use_pxe %}
+enable-tftp
+tftp-root=/var/lib/tftpboot/
```
I think you can use the `tftp_directory` variable

natifridman commented on roles/setup_tftp/handlers/main.yml:
```diff
@@ -0,0 +1,14 @@
+---
+- name: Restart TFTP
```
Do we need this file?

halleelyasaf replied:
not really

natifridman commented on playbooks/deploy_dns.yml:
```diff
@@ -2,6 +2,7 @@
   hosts: dns_host
   vars:
     SETUP_DNS_SERVICE: "{{ setup_dns_service | default(True) }}"
+    TFTP_ROOT: "{{ hostvars['tftp_host']['tftp_directory'] | default(/var/lib/tftpboot/) }}"
```
I think it should be in the TFTP role and no in the DNS

halleelyasaf replied:
It is two different roles one is for the dnsmasq and the other one is for the TFTP role

natifridman commented on roles/mount_discovery_iso_for_pxe/defaults/main.yml:
```diff
@@ -1,6 +1,6 @@
 ---
 # defaults file for generate_discovery_iso
-tftp_directory: "/var/lib/tftpboot"
+tftp_directory: "{{ hostvars['httpd_pxe_host']['httpd_port_for_pxe'] }}"
```
If you get the `TFTP_ROOT` var from the playbook, do you need this variable? You can use `TFTP_ROOT` instead

nocturnalastro committed:
Fixes issue with gather_facts in at least the ansible version used in lab

nocturnalastro committed:
Allow for users to store kubeadmin creds in plain text

In a lab enviroment encrypting the kubeadmin creds my be cumbersome
so if a user wishs they can not provide a
kubeadmin_vault_password_file_path and the user will be warned about
the behaviour in the deployment plan

natifridman commented on inventory.yml.sample:
```diff
@@ -216,10 +224,11 @@ all:
       #       then a "vm_host" must be defined in the node and a host with that name must exist
-      #       inside the "vm_hosts" group.
+      #       inside the "vm_hosts" group. 
+      #     - "PXE" identifies a node as a baremetal that need to boot from PXE. If a "PXE" node is present,
```
```suggestion
      #     - "PXE" identifies a node as a baremetal that needs to boot from PXE.
```

natifridman commented on README.md:
```diff
@@ -205,6 +205,12 @@ The following are defaults for a full setup:
   - The discovery image from Assisted Installer will be placed in and served from `/opt/http_store/data`
 
+- HTTP Store for pxe
+  - The discovery image from the httpd store will mount to this server and boot the rootfs
```
```suggestion
  - The discovery image from the httpd store will be mounted to this server and will boot the rootfs
```

natifridman commented on README.md:
```diff
@@ -205,6 +205,12 @@ The following are defaults for a full setup:
+  - The discovery image from the httpd store will mount to this server and boot the rootfs
+
+- TFTP Host
+  - The discovery image will mount to this server and do the pxe boot with TFTP 
```
```suggestion
  - The discovery image will be mounted to this server and do the PXE boot with TFTP 
```

snecklifter committed:
Moves Ansible boot override mode to redfish_command

With community.general 3.5.0, redfish_command supports overriding
the boot mode so lets use that rather than the uri module. This
should apply to other hardware vendors but I only have access to
ILO to test currently.

dependabot[bot] committed:
build(deps): bump actions/setup-python from 2.3.2 to 3

Bumps [actions/setup-python](https://github.com/actions/setup-python) from 2.3.2 to 3.
- [Release notes](https://github.com/actions/setup-python/releases)
- [Commits](https://github.com/actions/setup-python/compare/v2.3.2...v3)

---
updated-dependencies:
- dependency-name: actions/setup-python
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

natifridman commented on roles/mount_discovery_iso_for_pxe/defaults/main.yml:
```diff
@@ -1,17 +1,23 @@
 # defaults file for generate_discovery_iso
 tftp_root: "{{ hostvars['tftp_host']['tftp_directory'] | default('/var/lib/tftpboot/') }}"
-mount_directory: "/mnt/images"
+mount_directory: "/mnt/iso/images"
```
Maybe you can call it `mount_images_directory` and create another variable like `mount_efiboot_directory` so you can use it instead of using the full path in your playbook

natifridman commented on playbooks/README.md:
```diff
@@ -10,6 +10,9 @@ and are nominally put in the order of usage with some exceptions. This may be us
 | `deploy_http_store.yml`                | Deploys HTTP server                                           | -                     |
 | `deploy_registry.yml`                  | Deploys container image registry                              | -                     |
 | `deploy_dns.yml`                       | Deploys DNS and DHCP                                          | -                     |
+| `deploy_tftp.yml`                      | Deploys the tftp server                                          | -                     |
+| `deploy_httpd_for_pxe.yml`             | Deploys Httpd for the PXE                                          | -                     |
```
```suggestion
| `deploy_tftp.yml`                      | Deploys tftp server                                          | -                     |
| `deploy_httpd_for_pxe.yml`             | Deploys httpd for the PXE                                          | -                     |
```

natifridman commented on docs/inventory.md:
````diff
@@ -406,6 +409,51 @@ The basic network configuration of the inventory for the fully bare metal deploy
 ```
 
+## PXE Deployment
+this services need to set as true when using PXE deployment
````
```suggestion
These services are required when using PXE deployment
```

dependabot[bot] committed:
build(deps): bump actions/checkout from 2 to 3

Bumps [actions/checkout](https://github.com/actions/checkout) from 2 to 3.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v2...v3)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

abays committed:
Fix check for generate_ssh_keys

abays committed:
Ensure 'fetched' dir exists before use

nocturnalastro committed:
Add day2_cluster.txt file creation with day 2 cluster id

hhamid93 committed:
Fix qemu configuration so that non-root users can be used to create vms

nocturnalastro committed:
Add missing bracket in deployment plan

demvass committed:
Including day2 workers when inserting dns records

thekad committed:
Removing hard-coded sushy-tools path

You can set the `sushy_dir` variable, but the service file still pointed
to a static `/opt/sushy-tools` path.

dependabot[bot] committed:
build(deps): bump actions/cache from 2 to 3

Bumps [actions/cache](https://github.com/actions/cache) from 2 to 3.
- [Release notes](https://github.com/actions/cache/releases)
- [Commits](https://github.com/actions/cache/compare/v2...v3)

---
updated-dependencies:
- dependency-name: actions/cache
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

strigie committed:
Add jmespath to prereqs

snecklifter committed:
Fixes issues with port parsing

It seems there is an issue with later podman and its being
interpreted incorrectly. So we add quotes to ensure the colon
is part of the string.

p3ck committed:
Deploy virtual network before dns

If using the network bridge specified in the virtual environment the dns
role will reference it before it's been created which will cause dnsmasq
to fail to run.  This will break all dns lookups until fixed.

This PR deploys the virtual resources before the dns role.

thekad committed:
Install manual binaries in /usr/local/bin

One of the requirements says to install the `openshift-clients` package,
which provides `/usr/bin/{kubectl,oc}`. Then when populating the
registry we download the binaries according to the version to be
installed, then overwrite them on top of the ones in `/usr/bin`. It's
probably better to keep the integrity of the RPM DB and install them to
`/usr/local/bin` instead.

nocturnalastro committed:
Update style in Contributing

nocturnalastro committed:
Fix issue with setting day 2 hostname when not using dhcp

day 2 nodes were not getting there hostname set as the wait before
the patch did not include the insufficient state this meant when
using static ips the wait would time out. By addeding the state
and then adding another wait for ready after patching this issue
was fixed.

nocturnalastro committed:
Fix issue with SNO hostname not being set

While roles can not be set for SNO, the hostname can be so we build the patch
body from two tasks which create the correct body for SNO and for HA
